DirectProgramming/DPC++FPGA/Tutorials/Tools/dynamic_profiler/src/dynamic_profiler.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+  Please refer to the README file for information on how and why the
+  Intel(r) Dynamic Profiler for DPC++ should be used. This code sample 
+  does not explain the tool, rather it is an artificial example that
+  demonstates the sort of code changes the profiler data can guide.
+  The main content of this sample is in the README file.
+*/
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+
+// Two identical pipes to demonstrate the behaviour before
+// and after the design re-format
+using ProducerToConsumerBeforePipe =
+    INTEL::pipe<                           // Defined in the SYCL headers.
+        class ProducerConsumerBeforePipe,  // An identifier for the pipe.
+        float,                             // The type of data in the pipe.
+        20>;                               // The capacity of the pipe.
+using ProducerToConsumerAfterPipe =
+    INTEL::pipe<class ProducerConsumerAfterPipe, float, 20>;
+
+// Forward declare the kernel names
+// (This prevents unwanted name mangling in the optimization report.)
+class ProducerBeforeKernel;
+class ConsumerBeforeKernel;
+class ProducerAfterKernel;
+class ConsumerAfterKernel;
+
+// kSize = # of floats to process on each kernel execution.
+#if defined(FPGA_EMULATOR)
+constexpr int kSize = 4096;
+#else
+constexpr int kSize = 262144;
+#endif
+
+// Number of iterations performed in the consumer kernels
+// This controls the amount of work done by the Consumer.
+constexpr int kComplexity = 2000;
+
+// Perform two stages of processing on the input data.
+// The output of ConsumerWork1 needs to go to the input
+// of ConsumerWork2, so they cannot be done concurrently.
+// These functions are currently doing pointless work, but 
+// can be replaced with more useful operations.
+float ConsumerWork1(float f) {
+  float output = f;
+  for (int j = 0; j < kComplexity; j++) {
+    output = 20 * f + j - output;
+  }
+  return output;
+}
+
+float ConsumerWork2(float f) {
+  auto output = f;
+  for (int j = 0; j < kComplexity; j++) {
+    output = output + f * j;
+  }
+  return output;
+}
+
+/////////////////////////////////////////////////////////////
+// Pre-optimization kernel versions
+/////////////////////////////////////////////////////////////
+// The Producer kernel reads data from a SYCL buffer and writes it to
+// a pipe. This transfers the input data from the host to the Consumer kernel
+// that is running concurrently.
+// The Consumer kernel reads data from the pipe, performs the two ConsumerWork
+// operations on the data, and writes the results to the output buffer.
+
+void ProducerBefore(queue &q, buffer<float, 1>& buffer_a) {
+  auto e = q.submit([&](handler &h) {
+    // Get kernel access to the buffers
+    accessor a{buffer_a, h, read_only};
```

Comment:
For internal consistency, can you use round brackets for accessors?

Reply:
Done